KAFKA-15999: Migrate HeartbeatRequestManagerTest away from ConsumerTestBuilder

brenden20 opened the pull request:
In this PR I have completely migrated HeartbeatRequestManagerTest away from ConsumerTestBuilder. I have removed any instances of spy objects and used mocks wherever possible. 31/31 tests are passing. I also removed one test that existed in another test file.

brenden20 committed:
Revert "KAFKA-16713: Define initial set of RPCs for KIP-932 (#16022)"

This reverts commit 8f82f14a483d827d4653fb27b09945df936f49f6.

brenden20 committed:
Various test changes

brenden20 committed:
Stylistic changes

brenden20 committed:
Variable addition and cleanup

brenden20 committed:
Test updates

Updated testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() and testSuccessfulHeartbeatTiming() to get them to pass

brenden20 committed:
Removed testHeartbeatMetrics()

Removed testHeartbeatMetrics(), only test I was unable to get to pass. Also found that this test already exists in HeartbeatMetricsManagerTest

brenden20 committed:
Cleaning up

brenden20 committed:
Reapply "KAFKA-16713: Define initial set of RPCs for KIP-932 (#16022)"

This reverts commit d6bff1f19ae393c0dd94fb847bf2f74ef12fd08a.

brenden20 committed:
Cleaning up heartbeatRequestState vars

brenden20 committed:
Cleaning up membershipManager vars

brenden20 committed:
Cleaning up heartbeatRequestManager vars

brenden20 committed:
More cleanup

brenden20 committed:
Whitespace fixes

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 commented:
@philipnee @lianetm @kirktrue this is ready for pre-review. Let me know of any suggestions!

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -86,91 +79,169 @@
 
-        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
-    }
+        this.requestManagers = new RequestManagers(
```
is there a reason why we need this `RequestManagers`? I wouldn't expect to have it this test (it's a component above the HBReqManager even), and actually seems it's only created and closed?

brenden20 replied:
Actually no, that was an oversight on my part. Removed now.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -86,91 +79,169 @@
+        this.pollTimer = mock(Timer.class);
+        this.coordinatorRequestManager = mock(CoordinatorRequestManager.class);
+        this.heartbeatRequestState = mock(HeartbeatRequestState.class);
+        this.heartbeatState = mock(HeartbeatState.class);
```
turning this into a mock requires a little bit more work I would expect. Several tests assert on this request state (because it's mainly the internal component that dictates the timing of the heartbeats). Also I notice we end up having this mock and another request state that holds an actual instance (heartbeatRequestState1). So I wonder if this HeartbeatRequestState only is maybe a sensible usage for a spy in this test?

brenden20 replied:
Makes sense, I will remove ```heartbeatRequestState1``` and make ```heartbeatRequestState``` into a spy.

lianetm replied:
I notice now that my comment was misplaced here. I was talking about the `heartbeatRequestState` (and you addressed that, all good), but I placed it on the `heartbeatState`, which is another thing. 

So regarding the `heartbeatState`, that's the internal component of the HBMgr responsible of generating the heartbeat request content, and we do want to test that so I would say that we need it as an actual instance (not sure spy needed). One of the core tests for it is `testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments`  (see more comments on it).

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Remove requestManagers

brenden20 committed:
Merge branch '15999' of https://github.com/brenden20/kafka into 15999

brenden20 committed:
Changed heartbeatRequestState to spy

brenden20 committed:
Test cleanup

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'trunk' into 15999

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 committed:
Checkstyle fixes

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,99 @@ public class HeartbeatRequestManagerTest {
-        if (testBuilder != null) {
-            testBuilder.close();
-        }
+    private void resetWithZeroHeartbeatInterval() {
```
we can just remove this reset func now and call the `createHeartbeatStateWith0HeartbeatInterval` directly

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,99 @@ public class HeartbeatRequestManagerTest {
-        if (testBuilder != null) {
-            testBuilder.close();
-        }
+    private void resetWithZeroHeartbeatInterval() {
```
actually, this makes me notice. This reset just calls the create states but with 0 as interval. We already create the states in the beforeEach, so I expect we could simplify this and just `time.sleep(DEFAULT_HEARTBEAT_INTERVAL_MS)` in all the places where we call the `resetWithZeroHeartbeatInterval` ? we could then remove the reset and createHeartbeatStateWith0HeartbeatInterval I expect

brenden20 replied:
It works for all but one test, for now I am keeping ```createHeartbeatStateWith0HeartbeatInterval()``` for ```testSkippingHeartbeat(final boolean shouldSkipHeartbeat)``` since this test requires the initial heartbeatInterval be 0

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -194,27 +218,38 @@ public void testSuccessfulHeartbeatTiming() {
 
-        ConsumerGroupHeartbeatRequest heartbeatRequest =
-                (ConsumerGroupHeartbeatRequest) request.requestBuilder().build(version);
+        ConsumerGroupHeartbeatRequest heartbeatRequest = mock(ConsumerGroupHeartbeatRequest.class);
+        ConsumerGroupHeartbeatRequestData data = new ConsumerGroupHeartbeatRequestData();
+        data.setSubscribedTopicNames(Collections.singletonList(topic));
+        data.setRebalanceTimeoutMs(10000);
+        data.setGroupId("groupId");
+        data.setInstanceId("group-instance-id");
+
+        when(heartbeatRequest.data()).thenReturn(data);
```
uhm we shouldn't create the request data manually. We do want to test that the HBManager builds that data correctly from the state it has (the [buildRequestData](https://github.com/apache/kafka/blob/7d702e17604c6cc95e4e8bfdea58b04be3f5ffc3/clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java#L551) func inside the `heartbeatState`). So we should have the state as an actual instance and have the test as it was before I expect. See related [comment above](https://github.com/apache/kafka/pull/16200/files#r1649401668)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -194,27 +218,38 @@ public void testSuccessfulHeartbeatTiming() {
+    public void testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments() {
+        resetWithZeroHeartbeatInterval();
         String topic = "topic1";
         subscriptions.subscribe(Collections.singleton(topic), Optional.empty());
         membershipManager.onSubscriptionUpdated();
```
these 2 are mocks now right? so we should mock these calls if needed

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -275,29 +316,34 @@ public void testHeartbeatNotSentIfAnotherOneInFlight() {
+        // on the next poll waiting only for the minimal backoff.
+        inflightReq.handler().onComplete(createHeartbeatResponse(inflightReq, Errors.NONE));
+        time.sleep(DEFAULT_RETRY_BACKOFF_MS);
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
```
I wouldn't expect us to need to mock this canSend on the requestState that is a spy (the real method should kick in). But I guess you had to add it for some reason that was not generating the HB? I wonder if what we're missing is a time.sleep(HB_Interval) right before completing the response on ln 320? to ensure the interval has expired, the get a response, then check the next HB is sent only waiting for the backoff.

brenden20 replied:
Looks like at some point I changed the order, switched it back and it is working good now

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,59 +592,86 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
         data = heartbeatState.buildRequestData();
-        assertEquals(ConsumerTestBuilder.DEFAULT_GROUP_ID, data.groupId());
-        assertEquals(memberId, data.memberId());
+        data.setRebalanceTimeoutMs(10000);
+        data.setSubscribedTopicNames(Collections.singletonList(topic));
+        data.setTopicPartitions(Collections.emptyList());
+        data.setServerAssignor(DEFAULT_REMOTE_ASSIGNOR);
```
Similar to [this](https://github.com/apache/kafka/pull/16200/files#r1649402467) other comment. We shouldn't set the request content manually, because we want to check that the `buildRequestData` generates that content properly. I expect that when we turn the `heartbeatState` back into an instance, not a mock, this test could go back to how it was, maintaining the value it has checking the request content

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -659,78 +753,38 @@ public void testPollTimerExpirationShouldNotMarkMemberStaleIfMemberAlreadyLeavin
         mockStableMember();
 
         time.sleep(DEFAULT_HEARTBEAT_INTERVAL_MS);
+        when(membershipManager.isLeavingGroup()).thenReturn(true);
```
uhm do we need this here? I wouldn't expect so (the membershipMgr is a mock now, and the HB mgr does not check the isLeavingGroup to generate a HB)

lianetm replied:
also on ln 793 there's a doNothing().when(membershipManager) that I would say we don't need on the mock

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -744,34 +798,17 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
         verify(heartbeatRequestState).onFailedAttempt(anyLong());
         verify(heartbeatRequestState).reset();
 
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(false);
         when(membershipManager.state()).thenReturn(MemberState.FENCED);
```
what the HB manager checks on the membershipMgr in this case is the shouldSKip, so I guess what we need to do here is to mock the fencing is: `when(membershipManager.shouldSkipHeartbeat()).thenReturn(true);` (wonder if this will allow you to loose the when canSendRequest too, which I would expect should work because the heartbeatRequestState is not a mock)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -744,34 +798,17 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
         result = heartbeatRequestManager.poll(time.milliseconds());
         assertEquals(0, result.unsentRequests.size(), "Member should not send heartbeats while FENCED");
 
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
         when(membershipManager.state()).thenReturn(MemberState.JOINING);
```
ditto, we probably don't need to add an expectation on the canSend, and we just need to set the right one on the membershipMgr to show it's not fenced anymore (when(membershipManager.shouldSkipHeartbeat()).thenReturn(false))

brenden20 replied:
Good catch there, thank you!

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -659,78 +753,38 @@ public void testPollTimerExpirationShouldNotMarkMemberStaleIfMemberAlreadyLeavin
-    @Test
-    public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
+        when(membershipManager.state()).thenReturn(MemberState.STABLE);
```
setting this expectation is great, and it actually makes me notice that the `mockStableMember()` could probably be removed? it was just taking all the actions needed to put the membershipMgr into STABLE (when it was an instance, not a mock)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -744,34 +798,17 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
 
-    @ParameterizedTest
-    @ApiKeyVersionsSource(apiKey = ApiKeys.CONSUMER_GROUP_HEARTBEAT)
-    public void testSendingLeaveGroupHeartbeatWhenPreviousOneInFlight(final short version) {
```
uhm we shouldn't remove this test (it was actually an issue we discovered, fixed, and added the test for coverage). I see it involves checking the HB content (that is might be off just because the heartbeatState is a mock, but give it another try after addressing the comments above to turn that state into an actual instance able to build the HB content)

lianetm commented:
Hey @brenden20, I completed another pass, left some comments. Thanks!

brenden20 committed:
Cleanup and minor fixes

brenden20 committed:
Added test back and fixed

brenden20 committed:
Cleanup

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Fix testHeartbeatState()

brenden20 committed:
Fix testValidateConsumerGroupHeartbeatRequest()

brenden20 committed:
Merge branch '15999' of https://github.com/brenden20/kafka into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
-        setUp(ConsumerTestBuilder.createDefaultGroupInformation());
-    }
+        this.time = new MockTime();
+        Metrics metrics = new Metrics(time);
```
for aesthetic purpose, let's group all the `this` and non `this` i.e., can we move metrics to the bottom?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -277,27 +332,29 @@ public void testHeartbeatNotSentIfAnotherOneInFlight() {
         assertEquals(1, result.unsentRequests.size(), "A next heartbeat should be sent on " +
-            "the first poll after receiving a response that took longer than the interval, " +
-            "waiting only for the minimal backoff.");
+                "the first poll after receiving a response that took longer than the interval, " +
```
could you revert the changes here? (as well as the ones above and below) I think these are IDE indentation changes.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -594,24 +740,13 @@ public void testHeartbeatState() {
                 heartbeatState,
                 heartbeatRequestState,
                 backgroundEventHandler);
+
```
can we remove the extra space.

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Minor fixes

brenden20 committed:
Merge branch '15999' of https://github.com/brenden20/kafka into 15999

brenden20 commented:
@philipnee @lianetm thank you both for the feedback, I have addressed all comments

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
         when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
+        Map<Uuid, SortedSet<Integer>> map = new HashMap<>();
+        LocalAssignment local = new LocalAssignment(0, map);
+        when(membershipManager.currentAssignment()).thenReturn(local);
```
do we really need to set this expectation for all tests? this is only used when building the request if there were partitions reconciled, so I would expect to need it only in very specific tests around that. Worth double-checking.

brenden20 replied:
Removing it changed nothing, good catch!

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
-        mockStableMember();
+        time.sleep(DEFAULT_HEARTBEAT_INTERVAL_MS);
         assertEquals(0, heartbeatRequestManager.maximumTimeToWait(time.milliseconds()));
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(mock(Node.class)));
```
good point about mocking the `Node` for the coordinator, it's all we need. This made me realize we create real instances for it in many other places though, when we probably don't need to (ex. ln 141). Could you please try replacing them with mocks just like you did here?

brenden20 replied:
Done, also found and removed some unnecessary stubs for this

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
 
     @Test
     public void testHeartbeatOnStartup() {
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```
Not really hurtful but this is changing the path the test runs. Before this PR, the first poll on this test would return null because the member was unsubscribed (shouldSkipHeartbeat true). Now it's because there is no coordinator. Both go down the same path in the poll implementation but if we want to be true to what we had before, here we should mock there is a known coordinator, and when(membershipMgr.shouldSkipHB).thenReturn(true).

brenden20 replied:
I have made the change you suggested, let me know what you think.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,72 +102,90 @@ public class HeartbeatRequestManagerTest {
         assertEquals(1, result.unsentRequests.size());
 
         // Ensure we do not resend the request without the first request being completed
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```
this one is changing the semantics of the test. Before it used to validate that no HB is generated if a request is in-flight. But now it's validating that no HB is generated if the coordinator is unknown. We should probably ensure there is a coordinator and mock heartbeatRequestState.requestInFlight() true.

brenden20 replied:
Similar to above, I have made the change. Let me know if it looks good

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -251,23 +304,25 @@ public void testSkippingHeartbeat(final boolean shouldSkipHeartbeat) {
         assertEquals(DEFAULT_HEARTBEAT_INTERVAL_MS - 100, heartbeatRequestManager.maximumTimeToWait(time.milliseconds()));
 
         // Member in state where it should not send Heartbeat anymore
         when(subscriptions.hasAutoAssignedPartitions()).thenReturn(true);
```
I would dare to say that we shouldn't need this now that we don't have a spy on `membershipMgr` anymore (it was there were this is used for several assignment/callbacks checks, not at the HB mgr level)

brenden20 replied:
It is actually needed for this test. ```maximumTimeToWait()``` will always return 0 if ```pollTimer``` is a mock. Just did some testing turning ```pollTimer``` from a mock to a spy and it worked out pretty good. All tests are working, I think it may be beneficial to keep ```pollTimer``` a spy, since I notice in many places we ```verify``` on ```pollTimer```. Let me know what you think, my changes are pushed

brenden20 replied:
I can do this if I turn ```pollTimer``` into a spy, I tried that out and I think we should keep it. It makes the tests more strict on timing, since a mock would have to be stubbed to get the desired behavior. It also allows us to keep our ```verify()``` invocations on ```pollTimer```. Let me know what you think

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -251,23 +304,25 @@ public void testSkippingHeartbeat(final boolean shouldSkipHeartbeat) {
         // Member in state where it should not send Heartbeat anymore
         when(subscriptions.hasAutoAssignedPartitions()).thenReturn(true);
+        when(membershipManager.shouldSkipHeartbeat()).thenReturn(true);
         membershipManager.transitionToFatal();
```
we shouldn't need this explicit call on the mock now (it's replaced by the `shouldSkipHeartbeat` expectation you already set above).

brenden20 replied:
It seems that I do actually need it, I tried removing it and found that ```shouldSkipHeartbeat``` is not working properly without it. Even stubbing the method to return ```MemberState.FATAL``` does not work for some reason. This seems to be the only way to return an empty poll result

lianetm replied:
uhm that's weird. The only check that should happen there is the `shouldSkipHeartbeat`, and we're mocking it. We're doing the same in other tests (ex. `testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin`), where just mocking the `shouldSkipHeartbeat` leads to no HB as expected (no actual transition). Could you please double check, trying to remove just ln 298 again?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -277,12 +332,12 @@ public void testHeartbeatNotSentIfAnotherOneInFlight() {
 
         result = heartbeatRequestManager.poll(time.milliseconds());
```
we need to sleep the interval before polling here again. If not, we're not getting HB because the interval hasn't expired (not because of in-flight as we want to assert on the following line)

lianetm replied:
I see down below (ln 337-340) the sequence to validate that the interval expired and no HB is generated because of the inflight. I truly don't see the value in keeping this one then (ln 333-335), which is not related to in-flight at all (HB interval is tested in other funcs). Makes sense?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -298,6 +353,8 @@ public void testHeartbeatNotSentIfAnotherOneInFlight() {
         when(membershipManager.shouldSkipHeartbeat()).thenReturn(false);
         when(membershipManager.shouldHeartbeatNow()).thenReturn(true);
+        when(pollTimer.remainingMs()).thenReturn(Long.MAX_VALUE);
+        when(heartbeatRequestState.timeToNextHeartbeatMs(time.milliseconds())).thenReturn(1000L);
```
is this 1000 the interval? if so let's use the constant to make it clear.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -312,31 +369,33 @@ public void testHeartbeatOutsideInterval() {
         // Mimic network timeout
         result.unsentRequests.get(0).handler().onFailure(time.milliseconds(), new TimeoutException("timeout"));
 
+        time.sleep(1);
```
not needed right? the flow we had (and should maintain) is:
1. poll generates HB and receives timeout as response
2. check backoff enforced (sleep backoff -1 , no request)
3. check new HB after backoff (sleep the 1ms that was missing to have the full backoff and check HB generated).

Since we're mocking the membershipMgr now we should need to set expectations to ensure that we're not skipping the hb because membershipMgr.shouldSkipHb is true, and ensure that it's the backoff the one really kicking in.

brenden20 replied:
Yeah, removing it changes nothing with the test, removing.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -312,31 +369,33 @@ public void testHeartbeatOutsideInterval() {
+        assertEquals(1, result.unsentRequests.size());
+
         // Assure the manager will backoff on timeout
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```
this changes the semantics. This will achieve the result of not generating a HB, but we want to test that the HB is not generated because the backoff hasn't expired. So we should ensure there is a coordinator, ensure that the membershipMgr allows HB (shouldSKipHb false), and see that it's indeed the backoff check [here](https://github.com/apache/kafka/blob/a533e246e3ed5f6f6c5be4ebf9d29ae75cab557e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestState.java#L89) the one making that no HB is generated

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -312,31 +369,33 @@ public void testHeartbeatOutsideInterval() {
     }
 
     @Test
     public void testFailureOnFatalException() {
```
since we're here, could we improve the `verify(backgroundEventHandler).add(any())` to check it's an `ErrorEvent` instead of any? it's truly what we expect on a fatal response in HB

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -312,31 +369,33 @@ public void testHeartbeatOutsideInterval() {
-        mockStableMember();
+        time.sleep(DEFAULT_HEARTBEAT_INTERVAL_MS);
 
+        when(membershipManager.isLeavingGroup()).thenReturn(true);
```
do we need this? (it's for the expired poll timer flow, which is unrelated)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -312,31 +369,33 @@ public void testHeartbeatOutsideInterval() {
 
+        when(membershipManager.isLeavingGroup()).thenReturn(true);
         when(coordinatorRequestManager.coordinator()).thenReturn(Optional.of(new Node(1, "localhost", 9999)));
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
```
needed? the state is spy, so I would expect we get here the behaviour we want, based on the interval that we already expired above, all from the real instance

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -251,23 +304,25 @@ public void testSkippingHeartbeat(final boolean shouldSkipHeartbeat) {
         assertEquals(0, result.unsentRequests.size());
         assertEquals(DEFAULT_HEARTBEAT_INTERVAL_MS - 100, result.timeUntilNextPollMs);
+
+        when(pollTimer.remainingMs()).thenReturn(1800L);
```
`1800L` isn't entirely clear to me.

brenden20 replied:
That line is now removed

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -349,6 +408,10 @@ public void testNoCoordinator() {
 
+        when(pollTimer.isExpired()).thenReturn(false);
+        when(pollTimer.remainingMs()).thenReturn(Long.MAX_VALUE);
+        when(heartbeatRequestState.timeToNextHeartbeatMs(time.milliseconds())).thenReturn(1000L);
```
needed or wanted for some reason? again wondering because since the state is a spy on the internal component, I would expect to have the real implementation kicking in, based on the interval defined for the HBMgr.

brenden20 replied:
These are now removed

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -357,31 +420,62 @@ public void testNoCoordinator() {
 
         List<String> subscribedTopics = Collections.singletonList("topic");
-        subscriptions.subscribe(new HashSet<>(subscribedTopics), Optional.empty());
+        when(subscriptions.subscription()).thenReturn(Collections.singleton("topic"));
```
should we reference the same `subscribedTopics` var in the `thenReturn` as before?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +967,6 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
                 heartbeatRequestState,
                 backgroundEventHandler,
-                metrics);
+                new Metrics());
```
isn't metrics already initialized in the ctor?

brenden20 replied:
Yes, but if you use the same metrics for different objects it breaks everything

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -756,18 +846,18 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
 
-        membershipManager.leaveGroup();
-
+        when(heartbeatRequestState.canSendRequest(anyLong())).thenReturn(true);
```
we should `when(membershipManager.state()).thenReturn(MemberState.LEAVING)` instead I would say, to be true to the real flow. When a member is leaving the HB is generated because of the state it's in (and it also considers the canSend request, but with the state as a spy do we really need to set this expectation?)

brenden20 replied:
You're right, replacing that now, thank you

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -756,18 +846,18 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
         ConsumerGroupHeartbeatRequest heartbeatToLeave = getHeartbeatRequest(heartbeatRequestManager, version);
         assertEquals(ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH, heartbeatToLeave.data().memberEpoch());
 
+        when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```
similar to other comments above, this will achieve the result we want (no HB), but not for the reason we wanted to test. The intention of the test was to asssert that polling the HB mananager after the request to leave has been sent will generate no request (and that's because the member transitions to UNSUBSCRIBED on `onHeartbeatRequestSent`.  So probably here we should only set the expectation that shouldSkipHeartbeat returns true, which is what the HB mgr checks, and gets when the member is UNSUBSCRIBED. Makes sense?

brenden20 replied:
Yes I see, I appreciate the explanations on these as well, it definitely gives me a better understanding of how everything works together. Changed now

lianetm commented:
Hey @brenden20 , thanks for the updates! I completed another pass, left some comments. Thanks!

brenden20 committed:
Various text fixes/updates

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge remote-tracking branch 'origin/15999' into 15999

brenden20 committed:
PollTimer changed to spy

brenden20 committed:
PollTimer back to mock

brenden20 committed:
PollTimer change and test fixes

brenden20 committed:
Couple of test fixes

brenden20 committed:
Various test changes and cleanup

brenden20 commented:
@lianetm @philipnee thank you both for the feedback! I have addressed all comments and left some replies as well, let me know what you think, made some pretty major improvements over the last iteration.

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -201,8 +202,42 @@ public void testSuccessfulHeartbeatTiming() {
+                DEFAULT_GROUP_ID,
+                Optional.of(DEFAULT_GROUP_INSTANCE_ID),
+                0,
+                Optional.of(""),
```
won't affect the test, but just for the sake of correctness, if we don't want an assignor we should pass `Optional.empty()`.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -201,8 +202,42 @@ public void testSuccessfulHeartbeatTiming() {
+        createHeartbeatStateWithZeroHeartbeatInterval();
+        time.sleep(DEFAULT_HEARTBEAT_INTERVAL_MS);
         String topic = "topic1";
+        // Make a singleton set
+        HashSet<String> set = new HashSet<>();
+        set.add(topic);
```
indeed, let's make a `Collections.singleton` set (and remove the comment and reuse the var on ln 241 too)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -756,18 +801,19 @@ public void testFencedMemberStopHeartbeatUntilItReleasesAssignmentToRejoin() {
         ConsumerGroupHeartbeatRequest heartbeatToLeave = getHeartbeatRequest(heartbeatRequestManager, version);
         assertEquals(ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH, heartbeatToLeave.data().memberEpoch());
 
+        //when(coordinatorRequestManager.coordinator()).thenReturn(Optional.empty());
```
let's remove this

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -659,95 +749,50 @@ public void testPollTimerExpirationShouldNotMarkMemberStaleIfMemberAlreadyLeavin
     }
 
-    @Test
-    public void testHeartbeatMetrics() {
```
just to double check, we're removing this because it's covered in the `HeartbeatMetricsManangerTest` right?

brenden20 replied:
Yes, you are correct

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -201,8 +202,42 @@ public void testSuccessfulHeartbeatTiming() {
     @ApiKeyVersionsSource(apiKey = ApiKeys.CONSUMER_GROUP_HEARTBEAT)
     public void testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments(short version) {
-        resetWithZeroHeartbeatInterval(Optional.of(DEFAULT_GROUP_INSTANCE_ID));
+        membershipManager = new MembershipManagerImpl(
+                DEFAULT_GROUP_ID,
+                Optional.of(DEFAULT_GROUP_INSTANCE_ID),
+                0,
+                Optional.of(""),
+                subscriptions,
+                mock(CommitRequestManager.class),
+                (ConsumerMetadata) metadata,
+                logContext,
+                Optional.of(mock(ClientTelemetryReporter.class)),
+                backgroundEventHandler,
+                time,
+                new Metrics()
+        );
+        membershipManager.transitionToJoining();
```
actually I wonder if we really need an instance of the `membershipMgr` here (and all the explicit calls on it). Since we're testing how the HBMgr generates the data, we maybe only need to mock the `membershipMgr` funcs the HBMgr relies on, maybe something like this, and removing the membershipMgr instance and all the calls on it, that were only trying to get to the state where these values would be returned:
```
    private void mockJoiningMemberData() {
        when(membershipManager.memberId()).thenReturn("");
        when(membershipManager.memberEpoch()).thenReturn(0);
        when(membershipManager.groupId()).thenReturn(DEFAULT_GROUP_ID);
        when(membershipManager.groupInstanceId()).thenReturn(Optional.of(DEFAULT_GROUP_INSTANCE_ID));
        when(membershipManager.currentAssignment()).thenReturn(LocalAssignment.NONE);
    }
```

lianetm replied:
btw, if this suggestion works, please consider it for the other test on the heartbeat state, maybe we can simplify there too (`testHeartbeatState` and `testValidateConsumerGroupHeartbeatRequest`)

brenden20 replied:
I was able to get this working with this kind of approach, let me know what you think

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -425,6 +482,7 @@ topicId, mkSortedSet(0)
         assertEquals(Collections.singletonList(expectedTopicPartitions), heartbeatRequest1.data().topicPartitions());
 
         // Assignment did not change, so no assignment should be sent
```
nit: I noticed that the `getHeartbeatRequest` used here has a kind of copy of this comment, totally out of place (ln 497). Could you pls remove it?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -447,7 +505,8 @@ private ConsumerGroupHeartbeatRequest getHeartbeatRequest(HeartbeatRequestManage
     public void testHeartbeatResponseOnErrorHandling(final Errors error, final boolean isFatal) {
-        mockStableMember();
+        if (isFatal)
+            when(membershipManager.state()).thenReturn(MemberState.FATAL);
```
trying to understand why exactly we need this for, I ended up in the `ensureHeartbeatStopped` that asserts on the `membershipMgr` state to check it's fatal, was that the reason? I would say we should remove that assert on the mock that brings no value (and then probably this, if not required for anything else)

brenden20 replied:
Looks like the if statement was not necessary, the tests pass regardless. Removing the assert and stubbing still make the test work!

lianetm replied:
great, that's what I was expecting.

brenden20 committed:
Test updates

brenden20 committed:
Test fix

brenden20 committed:
Test fix

brenden20 committed:
Test fixes

brenden20 committed:
Test fix

brenden20 commented:
@lianetm thank you for the review, I have taken your suggestions and implemented them. Let me know if there is anything else.

brenden20 committed:
Update HeartbeatRequestManagerTest.java

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +854,42 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
+        when(membershipManager.currentAssignment()).thenReturn(LocalAssignment.NONE);
+        when(membershipManager.serverAssignor()).thenReturn(Optional.of("uniform"));
+        when(membershipManager.state()).thenReturn(MemberState.UNSUBSCRIBED);
```
this is a bit confusing because it's not what happens in practice. A member that is about to join (send first HB) is in JOINING state, not unsubscribed. I guess you needed this because of the existing `testHeartbeatState` that uses this func and expects the state to be UNSUBSCRIBED. I would say that's what is actually wrong. I would suggest we remove the `assertEquals(MemberState.UNSUBSCRIBED, membershipManager.state());` on ln 560, and this expectation here too. (we're only asserting on something we set ourselves)

brenden20 replied:
Makes sense, changed now

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +854,42 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+        else
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
+        when(membershipManager.serverAssignor()).thenReturn(Optional.of("uniform"));
+        when(membershipManager.state()).thenReturn(MemberState.STABLE);
```
Similar to previous comment. If this was needed to unblock asserts on `membershipMgr.state()`, we should remove those assertions and this, as it truly brings no value to test the expectation we set.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +854,42 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+        else
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
+        when(membershipManager.currentAssignment()).thenReturn(LocalAssignment.NONE);
+        when(membershipManager.serverAssignor()).thenReturn(Optional.of("uniform"));
```
should we use the constant we already have? DEFAULT_REMOTE_ASSIGNOR

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +854,42 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.of(DEFAULT_GROUP_INSTANCE_ID));
+        else
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
+        when(membershipManager.serverAssignor()).thenReturn(Optional.of("uniform"));
```
DEFAULT_REMOTE_ASSIGNOR ?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +854,42 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+        when(membershipManager.memberId()).thenReturn("");
+        when(membershipManager.memberEpoch()).thenReturn(0);
+        when(membershipManager.groupId()).thenReturn(DEFAULT_GROUP_ID);
+        if (instanceId)
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.of(DEFAULT_GROUP_INSTANCE_ID));
+        else
+            when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
```
what about simplifying this to:
```
 when(membershipManager.groupInstanceId()).thenReturn(Optional.ofNullable(instanceId));
```
Passing a String instanceId as param. I find it's also clearer to have the string param with the explicit `instanceId` to use (or null).

brenden20 replied:
I do like this implementation better, thank you!

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,61 +536,76 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
 
     @Test
     public void testHeartbeatState() {
+        CommitRequestManager commitRequestManager = mock(CommitRequestManager.class);
```
I would expect we don't need anything related to the commitRequestManager, because the `HBManager` we're testing knows nothing about it. It's the `MembershipMgr` the one that commits offsets before revoking partitions. Could you try removing this and the expectation on ln 571?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,61 +536,76 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
-        assertEquals(data.topicPartitions(), Collections.emptyList());
-        membershipManager.onHeartbeatRequestSent();
+        assertEquals(Collections.emptyList(), data.topicPartitions());
         assertEquals(MemberState.STABLE, membershipManager.state());
```
I think we should remove this now that the `membershipMgr` is a mock, as it should be :) (it has no value since it's passing only because we set the expectation ourselves on `mockDefaultMemberData`)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,61 +536,76 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
+        assertEquals(DEFAULT_REMOTE_ASSIGNOR, data.serverAssignor());
         assertEquals(Collections.emptyList(), data.topicPartitions());
-        membershipManager.onHeartbeatRequestSent();
         assertEquals(MemberState.JOINING, membershipManager.state());
```
similar, no value anymore. We should remove this (and the expectation we set with the value on `mockFencedToJoiningMemberData`)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,61 +536,76 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
+        assertEquals(DEFAULT_REMOTE_ASSIGNOR, data.serverAssignor());
         assertEquals(Collections.emptyList(), data.topicPartitions());
-        membershipManager.onHeartbeatRequestSent();
         assertEquals(MemberState.JOINING, membershipManager.state());
```
ditto

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -580,39 +617,26 @@ public void testHeartbeatState() {
+        mockReconcilingState();
 
         // We remain in RECONCILING state, as the assignment will be reconciled on the next poll
         assertEquals(MemberState.RECONCILING, membershipManager.state());
```
ditto (and the `mockReconcilingState` could be completely removed)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -514,61 +536,76 @@ private void assertNextHeartbeatTiming(long expectedTimeToNextHeartbeatMs) {
+        assertEquals(DEFAULT_REMOTE_ASSIGNOR, data.serverAssignor());
         assertEquals(Collections.emptyList(), data.topicPartitions());
-        membershipManager.onHeartbeatRequestSent();
         assertEquals(MemberState.UNSUBSCRIBED, membershipManager.state());
```
same as comment [here](https://github.com/apache/kafka/pull/16200/files#r1681562408)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -40,13 +38,13 @@
 import org.apache.kafka.common.requests.ConsumerGroupHeartbeatResponse;
 import org.apache.kafka.common.requests.RequestHeader;
 import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.common.telemetry.internals.ClientTelemetryReporter;
```
let's remove. This is making the build [fail](https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16200/59/pipeline/) with unused import

lianetm commented:
Thanks for the updates @brenden20 . This is looking very good, left some comments just to keep improving.

brenden20 committed:
Various improvements

brenden20 commented:
@lianetm thank you again for the review and helpful comments, I have implemented all suggestions!

brenden20 committed:
Merge branch 'apache:trunk' into 15999

brenden20 committed:
Update HeartbeatRequestManagerTest.java

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -621,7 +628,6 @@ public void testPollTimerExpiration() {
-        verify(heartbeatRequestState).reset();
         verify(membershipManager).onHeartbeatRequestSent();
 
         when(membershipManager.state()).thenReturn(MemberState.STALE);
```
maybe no needed anymore? (the following expectation on `shouldSkipHeartbeat` is probably all we need)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -659,95 +665,50 @@ public void testPollTimerExpirationShouldNotMarkMemberStaleIfMemberAlreadyLeavin
         verify(heartbeatRequestState).onFailedAttempt(anyLong());
         verify(heartbeatRequestState).reset();
 
+        when(membershipManager.shouldSkipHeartbeat()).thenReturn(true);
```
nice, but this probably means that the following line expecting FENCED can be removed.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -659,95 +665,50 @@ public void testPollTimerExpirationShouldNotMarkMemberStaleIfMemberAlreadyLeavin
         result = heartbeatRequestManager.poll(time.milliseconds());
         assertEquals(0, result.unsentRequests.size(), "Member should not send heartbeats while FENCED");
 
+        when(membershipManager.shouldSkipHeartbeat()).thenReturn(false);
```
same, maybe we can get rid of the following line setting state to JOINING?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +837,30 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+                new Metrics());
+    }
+
+    private void mockJoiningMemberData(String instanceId) {
```
I realize now that we have to properly set the expected state here because it is considered when building the request data, to determine if it should be a full heartbeat or not (HeartbeatRequestManager [ln#566](https://github.com/apache/kafka/blob/f595802cc752ed01dc74e9ab932209fe25a9d10b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java#L566)), so I would say that this should include `when(membershipManager.state()).thenReturn(MemberState.JOINING)`. The test is probably passing anyways because of the way this func is used (first HB where all previous sentFields are null), but just to be true to how this will happen in real-life scenarios.

brenden20 replied:
I see, good catch! Changed now

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +837,30 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+        when(membershipManager.serverAssignor()).thenReturn(Optional.of(DEFAULT_REMOTE_ASSIGNOR));
+    }
+
+    private void mockFencedToJoiningMemberData() {
```
nit: `mockRejoiningMemberData` is maybe better (fencing is just only one of the ways to get to rejoining, not specifically used here)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -898,6 +837,30 @@ private HeartbeatRequestManager createHeartbeatRequestManager(
+        when(membershipManager.groupInstanceId()).thenReturn(Optional.empty());
+    }
+
+    private void mockDefaultMemberData(String instanceId) {
```
nit: `mockStableMemberData` maybe?

lianetm commented:
Hey @brenden20 , thanks for the updates! Just a few minor comments left.

brenden20 committed:
Merge branch 'trunk' into 15999

brenden20 committed:
Cleanup

brenden20 committed:
Implementing suggestions

brenden20 commented:
@lianetm thank you for the review! I addressed the comments you left. Let me know if it is good

lianetm reviewed (approved):
Thanks @brenden20, great improvement! LGTM.

brenden20 committed:
Merge branch 'apache:trunk' into 15999

lianetm commented:
Hey @chia7712, would you maybe have some time to take a look at this? Nice improvement @brenden20 did on the HeartbeatRequestManager test. Thanks!

chia7712 commented:
I will take a look today!

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -360,57 +371,68 @@ public void testNoCoordinator() {
     @ParameterizedTest
     @ApiKeyVersionsSource(apiKey = ApiKeys.CONSUMER_GROUP_HEARTBEAT)
     public void testValidateConsumerGroupHeartbeatRequest(final short version) {
+        heartbeatState = new HeartbeatState(
```
I guess those creations can be moved to a helper to be shared by `testValidateConsumerGroupHeartbeatRequestAssignmentSentWhenLocalEpochChanges`, `testValidateConsumerGroupHeartbeatRequest` and `testFirstHeartbeatIncludesRequiredInfoToJoinGroupAndGetAssignments`

brenden20 replied:
Good point! I have moved those object creations into a helper method now. Let me know what you think.

chia7712 reviewed (commented):
@brenden20 thanks for this great patch. I have only one small comment for you. PTAL

brenden20 committed:
Moved some object creations into a helper method

brenden20 commented:
@chia7712 thank you for your feedback! I have addressed your comment by moving those object creations into a helper method. Let me know how it looks!

chia7712 commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:
```diff
@@ -106,60 +99,85 @@ public class HeartbeatRequestManagerTest {
-        if (testBuilder != null) {
-            testBuilder.close();
-        }
+    private void createHeartbeatStateandRequestManager() {
```
nit: `createHeartbeatStateandRequestManager` -> `createHeartbeatStateAndRequestManager`

brenden20 replied:
Whoops, missed that. Fixed now.

brenden20 committed:
Update HeartbeatRequestManagerTest.java

brenden20 commented:
@chia7712 fixed that small issue with the name. Let me know if there is anything else!

brenden20 committed:
Merge branch 'apache:trunk' into 15999
